Multi-language example

goto40 opened the pull request:
This is my first attempt to contribute to this project.
My motivation is to learn how to make multi-language projects (https://github.com/langium/langium/discussions/669).

This PR is about a new example in the langium examples-folder

* with two languages in one project,
* and cross language references. 

An overview of the use case / the languages is given in `examples/requirements-and-tests/README.md` in this PR.
There are also some TODOs of some things I would like to clarify in this review.

**I would be happy to get feedback on anything not conforming - or what needs to be reworked.**

The language can be tested after building

* interactively (F5 with the corresponding configuration) in the folder `examples/requirements-and-tests/example`.
* with the cli (a code generator generating a HTML summary): `./bin/cli generate-requirements-coverage example/requirements.req` from within the folder `examples/requirements-and-tests`
* with the unittests: `npm exec jest -t "requirements"` from within the root folder of the langium project. 

Note:
* The generated code is committed: is this the way it is intended? (we do not test the `langium:generate` if we do not call it).
* I made many micro commits - all of them could be stashed with a meaningful message - if this PR gets accepted after some rework...

spoenemann commented:
Thanks for the huge contribution! A multi-language example will definitely be useful to demonstrate how to do this and testing this functionality regularly.

A thought about the two languages: it would be nice if they would share one or more parser rules to demonstrate that you can define a common part of the syntax that is reused by multiple languages. Currently they only share terminal rules (and a cross-reference).

goto40 commented:
> A thought about the two languages: it would be nice if they would share one or more parser rules to demonstrate that you can define a common part of the syntax that is reused by multiple languages. Currently they only share terminal rules (and a cross-reference).

Thank you for the quick response of a first impression. Take your time - we should make a good and simple example to help people employ langium in complex multi-language scenarios!

**Concerning your comment:** What do you think of having a common rule to allow to document a contact person who is responsible for this requirements-file or tests-file?
```text
contact {
    user: "Mickey Mouse"
}
// ... (either requirements or tests are following, dependent of the type of file.)
```

More points (for later, after this PR):
* A point I would like to discuss later is the difference of referencing a model and using common parts/rules. For now (=for this PR) the current setup fulfills all requirements I have... 
* Also we could allow to reference to real test files (e.g. typescript or python files) which are validated to include the test names (as function or just as text). This would demonstrate how to reference an external (non-langium) model: in this case the file system)

spoenemann commented:
> What do you think of having a common rule to allow to document a contact person who is responsible for this requirements-file or tests-file?

Yes sounds good. I'd choose a more concise syntax without curly braces for the contact persons (avoid replicating JSON syntax).

Other ideas:
 - Define environments in which the system under test can be used (e.g. operating systems, product variants etc.). A requirement can specify for which environments it is valid, and a test can specify to which environments it is applied (should be a partial or complete subset of the requirement environments).
 - Define an expression language for logic predicates about parts of the system under test. Use this to formalize both the requirements as well as parts of the tests. This would be considerably more work, though.

goto40 commented:
> * Define an expression language for logic predicates about parts of the system under test. Use this to formalize both the requirements as well as parts of the tests. This would be considerably more work, though.

Yes, I think this will generate too much code: The goal is to demonstrate how multi language projects can look like (not to present a fully-featured requirements tool).

goto40 commented:
As suggested, I added the contact person and the environments (w/o any validation or code generation). This should make clear that we can share rules...

* If you think some **scope provider** would be helpful, I could add one. But I think the example is clear enough as it is to me and **a scope provider would not add any new information**.
* If you have suggestions how to **simplify** - I would be happy to rework parts of the example...
* I also added some docu to **clarify the goal of the example**: https://github.com/langium/langium/pull/676/commits/88b90cc6f724e490be79f016ee95eaacb8a9b9b3

There are also some TODO in the README I would like to discuss (see inline comments).

goto40 commented on examples/requirements-and-tests/README.md:
```diff
@@ -0,0 +1,71 @@
+
+Language configuration:
+
+* `langium-config.json`: TODO, what does this exactly (AST code generation, syntax highlighting, TextMate).
```
@spoenemann Can we clarify this?

spoenemann replied:
The main purpose of that config file is to specify what languages to process when running the `langium` CLI. You can simply point to https://www.npmjs.com/package/langium-cli for documentation.

goto40 commented on examples/requirements-and-tests/README.md:
```diff
@@ -0,0 +1,71 @@
+Language configuration:
+
+* `langium-config.json`: TODO, what does this exactly (AST code generation, syntax highlighting, TextMate).
+* `package.json`: `contributes` section, TODO, what does this exactly (vscode extension linkage?).
```
@spoenemann Can we clarify this?

spoenemann replied:
You can point to https://code.visualstudio.com/api/references/contribution-points#contributes.languages for documentation.

spoenemann commented on examples/requirements-and-tests/README.md:
```diff
@@ -0,0 +1,71 @@
+
+* A test identifier shall contain a number.
+* A requirement identifier shall contain a number.
+* A requirement shall be covered by at least one test (TODO: when updating the test model, the requirement is not rechecked automatically).
```
The third point is hard to validate locally because you'd need to scan the whole workspace to find tests that cover a requirement. Typically this would rather be done as additional validation in a CLI or other processing tool, where you know the whole context.

Other idea for validation: a test must specify only environments that are mentioned in the corresponding requirement.

goto40 commented:
Note: I had to downgrade jest in my branch because of https://github.com/mattphillips/jest-chain/issues/34
This will be reverted when if the branch is merged into main (since the problem will disappear in the future  and the problem is only relevant if certain tests are failing).

spoenemann commented on examples/requirements-and-tests/src/language-server/requirements-language-module.ts:
```diff
@@ -0,0 +1,29 @@
+ * declared custom services. The Langium defaults can be partially specified to override only
+ * selected services, while the custom services must be fully specified.
+ */
+ export const RequirementsLanguageModule: Module<RequirementsLanguageServices, PartialLangiumServices & RequirementsLanguageAddedServices> = {
```
These declarations could be moved to requirements-and-tests-language-module.ts

spoenemann commented on examples/requirements-and-tests/src/language-server/tests-language-module.ts:
```diff
@@ -0,0 +1,29 @@
+ * declared custom services. The Langium defaults can be partially specified to override only
+ * selected services, while the custom services must be fully specified.
+ */
+export const TestsLanguageModule: Module<TestsLanguageServices, PartialLangiumServices & TestsLanguageAddedServices> = {
```
These declarations could be moved to requirements-and-tests-language-module.ts

spoenemann commented on examples/requirements-and-tests/package.json:
```diff
@@ -0,0 +1,74 @@
+        "@typescript-eslint/parser": "^5.28.0",
+        "eslint": "^8.17.0",
+        "langium-cli": "^0.4.0",
+        "typescript": "^4.7.4"
```
Some of these dependencies are already declared in the root package.json, please remove those from here to reduce maintenance effort.

spoenemann reviewed (commented):
Thanks for all this effort! I made suggestions to clean up the new project. Could you also rebase your branch on the current `main`? The PR shows a few changes that seem unrelated to your contribution.

goto40 commented:
> Thanks for all this effort! I made suggestions to clean up the new project. Could you also rebase your branch on the current `main`? The PR shows a few changes that seem unrelated to your contribution.

Thank you very much for the detailed answer! I will work on it the next days.

goto40 commented:
I rebased the branch and incorporated most comments.

Open points, please see inline comments for the following files above
* `examples/requirements-and-tests/src/language-server/requirements-language-module.ts`
* `examples/requirements-and-tests/src/language-server/tests-language-module.ts`

Please let me know about any points which are not ok.

spoenemann commented on examples/requirements/src/cli/index.ts:
```diff
@@ -0,0 +1,42 @@
+
+    const fileExtensions = RequirementsLanguageMetaData.fileExtensions.join(', ');
+    program
+        .command('generate-requirements-coverage')
```
The command name is a bit long, but we can improve this later...

spoenemann reviewed (approved):
Great work, thank you!

spoenemann commented:
@goto40 please resolve the eslint errors (`npm run lint`), then we can merge this.

goto40 commented:
> @goto40 please resolve the eslint errors (`npm run lint`), then we can merge this.

done. I have overseen this...
